src/thread/thread.cpp

```diff
@@ -127,8 +128,10 @@ namespace fc {
    }
 
    void thread::cleanup() {
-     delete current_thread();
-     current_thread() = nullptr;
+     if ( current_thread() ) {
+        delete current_thread();
+        current_thread() = nullptr;
+     }
```

Comment:
IIRC someone said it's fine to `delete nullptr` in C++11.

Reply:
You're correct. Yet I would question whether it is a good practice. At least it would not pop out in a valgrind output.

But the underlying pointer variable is static and can be accessed from various instances, at least theoretically. It is not easily seen that an instance won't manipulate the variable when it will point to an already deleted object, thus the more important change is clearing the variable. Now, there is another change within this commit. Together it could avoid loosing a not deleted object.

Honestly, I have not tested this particular change with differentiating the valgrind outputs, it takes a lot of time and the changes shall not do any harm.

Reply:
I don't think this change will have an effect, because there is nothing to clean up where you call `cleanup`. `current_thread()` returns a thread-local pointer within a thread that has just been created.

It probably won't hurt either.

Reply:
It is possible it won't have any effect. Maybe we could try to deduce something with Your better code-base knowledge.

What I can easily see. The `current_thread()`, which is actually just a static raw pointer, holds the most recently created instance of `class thread`. The only place where the `current_thread()` pointer is released is in the `cleanup()` method. Not counting the here-added call of this method, I can find only one place where it is called (asio.cpp::134).

If a new thread is created and then destroyed, won't there remain an invalid pointer? Are thread objects always dealt with properly? I.e., are they always destroyed, by the creator or by the usage pattern?

---

I don't remember the original loss record from valgrind which led me to this change, but meanwhile I have found out another one where I can clearly see that another problem resides there.
```
152 bytes in 1 blocks are still reachable in loss record 30 of 67
   at 0x502317F: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
   by 0x3007A01: fc::thread::wait_until(fc::shared_ptr<fc::promise_base>&&, fc::time_point const&) (thread.cpp:392)
   by 0x301AC76: fc::promise_base::_wait_until(fc::time_point const&) (future.cpp:89)
   by 0x301AAFC: fc::promise_base::_wait(fc::microseconds const&) (future.cpp:52)
   by 0x2AC8178: fc::promise<void>::wait(fc::microseconds const&) (future.hpp:152)
   by 0x30052DC: fc::thread::thread(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (thread.cpp:92)
   by 0x330A2A3: void __gnu_cxx::new_allocator<fc::thread>::construct<fc::thread, char const (&) [4]>(fc::thread*, char const (&) [4]) (new_allocator.h:136)
   by 0x33016DB: void std::allocator_traits<std::allocator<fc::thread> >::construct<fc::thread, char const (&) [4]>(std::allocator<fc::thread>&, fc::thread*, char const (&) [4]) (alloc_traits.h:475)
   by 0x32DA76D: std::_Sp_counted_ptr_inplace<fc::thread, std::allocator<fc::thread>, (__gnu_cxx::_Lock_policy)2>::_Sp_counted_ptr_inplace<char const (&) [4]>(std::allocator<fc::thread>, char const (&) [4]) (shared_ptr_base.h:526)
   by 0x32CB028: std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<fc::thread, std::allocator<fc::thread>, char const (&) [4]>(std::_Sp_make_shared_tag, fc::thread*, std::allocator<fc::thread> const&, char const (&) [4]) (shared_ptr_base.h:637)
   by 0x32B9CBF: std::__shared_ptr<fc::thread, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<std::allocator<fc::thread>, char const (&) [4]>(std::_Sp_make_shared_tag, std::allocator<fc::thread> const&, char const (&) [4]) (shared_ptr_base.h:1295)
   by 0x32A6996: std::shared_ptr<fc::thread>::shared_ptr<std::allocator<fc::thread>, char const (&) [4]>(std::_Sp_make_shared_tag, std::allocator<fc::thread> const&, char const (&) [4]) (shared_ptr.h:344)
```

I will investigate.

PS: The line numbers may not match the bitshares head revision.

Reply:
> The current_thread(), which is actually just a static raw pointer, 
> If a new thread is created and then destroyed, won't there remain an invalid pointer?

Yes and no. Mostly no, in my understanding. :-)
It's not just a static pointer, it's a `static __thread thread*`. The `__thread` means that this static variable has a different value for each thread. So yes, if a thread is created and then destroyed the pointer will be invalid, but the only thread that could have accessed it doesn't exist anymore.

I'm not sure how these `__thread` variables are implemented under the hood, and if valgrind understands them. It is of course possible that valgrind still sees the inaccessible variable and complains that it points to an invalid location.

Reply:
I missed this keyword. Thank You for enlightening me. :)

This makes my complaints yet less serious. Still, I would suggest accepting this change for the sake of easing analysing the code with valgrind. For anybody.

My investigation has diverged quite a lot and with questionable outcomes. I have not finished it yet, but I am taking a break and will possibly open up other PRs.